lib/staggered_prolong_restrict.cu

```diff
@@ -1,216 +1,54 @@
-    int x_cb = blockIdx.x*blockDim.x + threadIdx.x;
-    int parity = arg.nParity == 2 ? blockDim.y*blockIdx.y + threadIdx.y : arg.parity;
-    if (x_cb >= arg.fineVolumeCB) return;
-
-    int c = blockDim.z*blockIdx.z + threadIdx.z;
-    if (c >= fineColor) return;
-
-    staggeredProlongRestrict<Arg::transferType>(arg.out, arg.in, parity, x_cb, c, arg.geo_map, arg.spin_map, arg.fineVolumeCB, arg.coarseVolumeCB, arg.fineX);
-  }
-  
   template <typename Float, int fineSpin, int fineColor, int coarseSpin, int coarseColor, StaggeredTransferType transferType>
-  class StaggeredProlongRestrictLaunch : public TunableVectorYZ {
+  class StaggeredProlongRestrictLaunch : public TunableKernel3D {
+    template <QudaFieldOrder order> using Arg = StaggeredProlongRestrictArg<Float,fineSpin,fineColor,coarseSpin,coarseColor,order,transferType>;
 
     ColorSpinorField &out;
     const ColorSpinorField &in;
     const int *fine_to_coarse;
     int parity;
-    QudaFieldLocation location;
-    char vol[TuneKey::volume_n];
 
-    bool tuneGridDim() const { return false; } // Don't tune the grid dimensions.
     unsigned int minThreads() const { return fineColorSpinorField<transferType>(in,out).VolumeCB(); } // fine parity is the block y dimension
 
   public:
     StaggeredProlongRestrictLaunch(ColorSpinorField &out, const ColorSpinorField &in,
-                     const int *fine_to_coarse, int parity)
-      : TunableVectorYZ(fineColorSpinorField<transferType>(in,out).SiteSubset(), fineColor), out(out), in(in),
-        fine_to_coarse(fine_to_coarse), parity(parity), location(checkLocation(out, in))
+                                   const int *fine_to_coarse, int parity) :
+      TunableKernel3D(fineColorSpinorField<transferType>(in,out), fineColorSpinorField<transferType>(in,out).SiteSubset(), fineColor),
+      out(out),
+      in(in),
+      fine_to_coarse(fine_to_coarse),
+      parity(parity)
     {
-      strcpy(vol, fineColorSpinorField<transferType>(in,out).VolString());
       strcat(vol, ",");
       strcat(vol, coarseColorSpinorField<transferType>(in,out).VolString());
-
-      strcpy(aux, fineColorSpinorField<transferType>(in,out).AuxString());
       strcat(aux, ",");
       strcat(aux, coarseColorSpinorField<transferType>(in,out).AuxString());
+
+      apply(device::get_default_stream());
     }
 
     void apply(const qudaStream_t &stream) {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
       if (location == QUDA_CPU_FIELD_LOCATION) {
         if (out.FieldOrder() == QUDA_SPACE_SPIN_COLOR_FIELD_ORDER) {
-          StaggeredProlongRestrictArg<Float,fineSpin,fineColor,coarseSpin,coarseColor,QUDA_SPACE_SPIN_COLOR_FIELD_ORDER,transferType>
-            arg(out, in, fine_to_coarse, parity);
-          StaggeredProlongRestrict<Float,fineSpin,fineColor,coarseSpin,coarseColor>(arg);
+          launch_host<StaggeredProlongRestrict_>(tp, stream, Arg<QUDA_SPACE_SPIN_COLOR_FIELD_ORDER>(out, in, fine_to_coarse, parity));
```

Comment:
No underscore at end of `StaggeredProlongRestrict_`

Reply:
83ec85e5ffb9bccc66ad43ed832d912433d1aa52